File: scala/strain/src/main/scala/Strain.scala
import scala.collection.mutable.ListBuffer

object Strain {
  def discard[T](list: Seq[T], clo: T => Boolean) = {
    list.foldLeft(ListBuffer[T]()) {
      (acc, n) => if (!clo(n)) acc += n else acc
    }
  }

  def keep[T](list: Seq[T], clo: T => Boolean) = {
    list.foldLeft(ListBuffer[T]()) {
      (acc, n) => if (clo(n)) acc += n else acc
    }
  }

}

File: ruby/strain/strain.rb
# frozen_string_literal: true

class Array
  def keep(&clo)
    reduce([]) { clo.call(_2) ? _1 << _2 : _1 }
  end

  def discard(&clo)
    reduce([]) { clo.call(_2) ? _1 : _1 << _2 }
  end
end
